cpp/arcticdb/util/reliable_storage_lock.hpp

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#pragma once
+
+#include <arcticdb/storage/store.hpp>
+#include <arcticdb/entity/types.hpp>
+#include <arcticdb/util/constants.hpp>
+#include <arcticdb/util/configs_map.hpp>
+
+#include <mutex>
+#include <folly/experimental/FunctionScheduler.h>
+
+namespace arcticdb {
+
+namespace lock {
+
+using Epoch = uint64_t;
+
+// The ReliableStorageLock is a storage lock which relies on atomic If-None-Match Put and ListObject operations to
+// provide a slower but more reliable lock than the StorageLock. It should be completely consistent unless a process
+// holding a lock get's paused for times comparable to the lock timeout.
+// It lock follows the algorithm described here:
+// https://www.morling.dev/blog/leader-election-with-s3-conditional-writes/
```

Comment:
Maybe we should save of copy of this blog somewhere, will suck to have no docs of how this works if it vanishes

Reply:
Given that we'll soon replace the locking with If-Match based locks, we can probably skip this.